script/create_changelog

```diff
@@ -0,0 +1,250 @@
+	echo "   This script doesn't print your USERNAME on output"
+	echo ""
+	echo -e "- \033[1mTOKEN\033[0m"
+	echo "   This is github token. This script doesn't print your TOKEN on output."
+	echo "   To create a new token,"
+	echo "   refer https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line"
+	echo ""
+	echo -e "- \033[1mEMAIL\033[0m"
+	echo -e "   This is your email id, configured through \033[3mgit config --global user.email\033[0m, to sign the commit"
+	echo "   This script doesn't print your EMAIL on output"
+	echo ""
+	echo -e "\033[1mHow this script works?\033[0m"
+	echo -e "- \033[3mGenerating changelog\033[0m"
+	echo -e "   Script will use files from directory \033[2mchangelogs/unreleased\033[0m to create a changelog"
+	echo -e "   It will create a new directory, named \033[4mtag-name\033[0m, under directory \033[2mchangelogs/released\033[0m"
+	echo -e "   and move all the files from directory \033[2mchangelogs/unreleased\033[0m to new directory."
+	echo "   Once all the files are moved to new directory, it will create a temporary file "
+	echo -e "   in \033[2m/tmp/\033[0m directory and copy the content of all files, under new directory to a"
+	echo "   temporary file."
+	echo -e "   If tag is release tag, like \033[2mv1.1.0\033[0m, and there are existing directory for rc release,"
+	echo -e "   like \033[2mv1.1.0-RC1\033[0m, then script will copy the contents of all files, under this rc release "
+	echo "   directory, to a temporary file."
+	echo "   After this, script will add the content of temporary file to root changelog file "
+	echo "   and delete the temporary file"
+	echo ""
+	echo -e "- \033[3mCreating a branch\033[0m"
+	echo -e "  Script creates a new branch named \033[2mchangelog-timestamp\033[0m, like \033[2mchangelog-20200512164456\033[0m"
+	echo ""
+	echo -e "- \033[3mCreating a commit\033[0m"
+	echo -e "   Script will create commit with following files using your \033[1mUSERNAME\033[0m and \033[1mEMAIL\033[0m"
+	echo -e "   - \033[2mrepo-name/CHANGELOG.md\033[0m (\033[3mmodified\033[0m)"
+	echo -e "   - \033[2mrepo-name/changelogs/unreleased/*\033[0m (\033[3mdeleted\033[0m)"
+	echo -e "   - \033[2mrepo-name/changelogs/released/tag-name\033[0m (\033[3mnew directory\033[0m)"
+	echo ""
+	echo -e "- \033[3mPushing branch to repo\033[0m"
+	echo -e "   To push the changes to github, script will add new remote, named \033[2mupstream\033[0m, "
+	echo -e "   using your \033[1mUSERNAME\033[0m and \033[1mTOKEN\033[0m"
+	echo ""
+	echo -e "- \033[3mCreating a pull request\033[0m"
+	echo -e "   Script uses github API to create a pull request, using your \033[1mUSERNAME\033[0m and \033[1mTOKEN\033[0m."
+	echo "   Refer https://developer.github.com/v3/pulls/#create-a-pull-request for API"
+	exit 1
+fi
+
+if [ -z $USERNAME ] || [ -z $TOKEN ] || [ -z $EMAIL ]; then
+	echo "No crendentials provided for github"
+	exit 1
+fi
+
+BRANCH=changelog-$(date +'%Y%m%d%H%M%S')
```

Comment:
@mynktl -- would it be possible to move the change-log generate and rasing PR into different scripts. 

thinking of the following usecase:
- we have not yet fully mastered the best practicies around - writting good comments or organizing the fixes. this may take one or two more releases. 
- have an option for the repo maintainer to locally run the script, check the git diff, make the necessary modifications and decide to raise the PR.

Reply:
One practice - probably we need to adopt some of the comments addressed in this PR about changelogs appearing in master as well as release branch could be: 
- decide that it is ok for a PR to be present in both changelogs of current release - and next release
- don't add a changelog to master, if it is going to be cherry-picked (maybe change the practice to first fix in the release branch and then cherry-pick to master)

Reply:
ok. I will fragment this script into two script -- one to create changelog branch and create commit and second to raise a PR.
> decide that it is ok for a PR to be present in both changelogs of current release - and next release

changelog will appear on master branch only. If someone wants to read changelog, He/She will open repo page only and that will redirect it to default branch.
> don't add a changelog to master, if it is going to be cherry-picked (maybe change the practice to first fix in the release branch and then cherry-pick to master)

This is a very rare scenario since. In this scenario owner can update the branch itself, used in PR.
> we have not yet fully mastered the best practicies around - writting good comments or organizing the fixes. this may 
take one or two more releases.

Agree. Writing good comments, I've added comments in script where i thought necessary. But it may depends on other users according to their knowledge of bash. It may vary. If i've not added comment on any complex logic which is difficult to understand, please drop a comment.
> have an option for the repo maintainer to locally run the script, check the git diff, make the necessary modifications and decide to raise the PR.

repo maintainer can run this script locally and update the PR. Fragmenting this script into two script -- to create commit and to raise a PR -- will solve this.

Reply:
can it be split as: 
- make changes 
- commit and raise PR. 

Once the script commits, it includes modifying and adding a new commit or squash related steps. 

re: Comments, I meant overall contributions to the repo, not just on this PR. They may be coming from new contributors at times, who have not yet fully understood the use-case or guidelines. Of course, even the most experienced folks cannot always get the comments right, as english doesn't have to be contributors' native language.

Reply:
> -   make changes
> -  commit and raise PR.


I'm good with above approach. Thinking of adding some option in the script itself. Reason of this script is to automate the whole changelog process. If we split the script then maintainer need to do some extra work to execute second script.

Reply:
About comments, PR review surely helps here, If comments are not placed(for complex code or critical section which requires additional measures for future changes) or not proper.

Reply:
@kmova I've updated the script to generate changelog only and also if user want to raise a PR against release or production branch -- he/she can do that if arguments,with value, are mentioned.